backend/dataall/modules/dataset_sharing/services/share_processors/lf_process_cross_account_share.py

```diff
@@ -157,10 +157,23 @@ def process_revoked_shares(self) -> bool:
 
                 self.revoke_table_resource_link_access(table, principals)
 
+                other_table_shares_in_env = False
+                if ShareObjectRepository.other_approved_share_item_table_exists(
+                    self.session,
+                    self.target_environment.environmentUri,
+                    share_item.itemUri,
+                    share_item.shareItemUri
+                ):
+                    other_table_shares_in_env = True
+                    principals = [p for p in principals if "arn:aws:quicksight" not in p]
```

Comment:
I see now the issue that you commented about the Quicksight groups. You are right, there is one single QS group for the whole environment. Which means that removing permissions for the QS group should happen when we efectuate the "clean_up". The clean-up consists of deleting the sahred database, and I think by deleting the shared database the existing Lakeformation permissions also get deleted right? In that case we already got what we want.

Reply:
For Each Revoked Table:
- Revoke Table Resource Link Access (ALL Principals)
- Revoke Source Table Access (ALL Principals)
- Delete Resource Link Table
- If no other existing shared Table Items in the share
    - Delete Shared DB
    - If no other approved share objects for dataset (i.e. `other_approved_share_object_exists`)
        - Revoke External Account Access

The Problems with the above:
- The method `other_approved_share_object_exists()` was filtering on Share Object Status of Approved (not getting all shared states so filtering incorrectly)
    - Meaning we were revoking target account access when there were still other shares for that dataset

- Could be prematurely removing QS Permissions (since 1 QS Group used for all teams in the env)
    - Could have QS Permissions revoked to tableA because 1 share revoked but another for a different team in same environment still exists and needs those QS Permissions

- If there was a share of 2 tables in 1 dataset to a team that were revoked at separate times
    - Since we iterate on `self.revoked_tables` to revoke external target account access to each table if no other shares for that dataset exist we would only catch the last table that is included in the self.revoked_table and may miss on some permissions to clean up
    - We are doing dataset-level share object exist checks but revoking external access on table-level


Now:
- Revoke Table Resource Link Access (ALL Principals)
- Check if the same table is currently shared to other teams in environment
    - If true revoke source table access only to IAM role (NOT QS if present)
    - If false revoke ALL principals (QS included) and revoke account access on the table-level
- Delete Resource Link Table
- If no other existing shared Table Items
    - Delete Shared DB

Fixes
- Fixes `other_approved_share_object_exists()` and makes check on table level to properly check if other share for that table item in that dataset exist on the target environment

- Does not remove QS Permissions on LakeFormation if there are other shares for that table in that environment

- Revokes external target account access on a table-level if there are no other shares on the table item in the target account
    - Table-level checks and table-level revokes on permissions